### PR TITLE
AF-1100 : [Project Oriented] After project is renamed, its breadcrumb and title are not updated

### PR DESCRIPTION
**Pull request by Rikkola**
https://issues.jboss.org/browse/AF-1100

https://github.com/kiegroup/appformer/pull/379
https://github.com/kiegroup/kie-wb-common/pull/1846

### CODE REVIEW

**Comment by Rikkola** [comment]
@ederign @jhrcek I would appreciate if you could also take a look at the kie-wb-common PR.

**Comment by jhrcek** [comment]
@Rikkola sure, will take a look. 
Jenkins execute full downstream build

File: uberfire-project/uberfire-project-backend/src/main/java/org/guvnor/common/services/project/backend/server/POMServiceImpl.java [hunk 1]
```diff
@@ -153,74 +163,118 @@ public Path save(final Path path,
...
+
+            if (content.isMultiModule() &&
+                    content.getModules() != null) {
+
+                for (final String module : content.getModules()) {
+
+                    final org.uberfire.java.nio.file.Path childPath = Paths.convert(path).getParent().resolve(module).resolve("pom.xml");
+
+                    if (ioService.exists(childPath)) {
+                        final POM child = load(Paths.convert(childPath));
```
Comment (jhrcek) [inline uberfire-project/uberfire-project-backend/src/main/java/org/guvnor/common/services/project/backend/server/POMServiceImpl.java]: I'd rename `child` -> `childContent` to keep parameter naming consistency.

File: uberfire-project/uberfire-project-backend/src/main/java/org/guvnor/common/services/project/backend/server/POMServiceImpl.java [hunk 2]
```diff
@@ -153,74 +163,118 @@ public Path save(final Path path,
...
+                        if (child != null) {
+                            child.setParent(content.getGav());
```
Comment (jhrcek) [inline uberfire-project/uberfire-project-backend/src/main/java/org/guvnor/common/services/project/backend/server/POMServiceImpl.java]: Id save ` GAV parentGav = content.getGav();` for clarity and not to repeat the call 3 times. Also looking at how deep this is nested, maybe extracting this out to helper would help readability too

```
content.getModules()
    .forEach(childModule -> saveGavChange(parentGav, childModule));

```

File: uberfire-project/uberfire-project-backend/src/main/java/org/guvnor/common/services/project/backend/server/POMServiceImpl.java [hunk 3]
```diff
@@ -153,74 +163,118 @@ public Path save(final Path path,
...
-        } else {
-            ioService.write(Paths.convert(path),
-                            pomContentHandler.toString(content,
-                                                       loadPomXMLString(path)),
-                            metadataService.setUpAttributes(path,
-                                                            metadata));
+        private Optional<Module> getModuleIfNameChanged(final Path path,
+                                                        final POM content) {
+            if (!load(path).getName().equals(content.getName())) {
+                return Optional.of(moduleService.resolveModule(path));
```
Comment (jhrcek) [inline uberfire-project/uberfire-project-backend/src/main/java/org/guvnor/common/services/project/backend/server/POMServiceImpl.java]: This seems incorrect to me (or I simply misunderstand how it works). You're calling `moduleService.resolveModule(path)` here and returning it in Optional. And then on line 241 you're creating RenameModuleEvent **with the same path** `moduleService.resolveModule(path)` (the path didn't change along the way). Is there some magic involved or are you actually passing the same 'old module' twice to RenameModuleEvent constructor?

Also renaming all occurrences of `path` to `pomPath` in this entire class (where it applies) would improve readability..

Comment (Rikkola) [reply uberfire-project/uberfire-project-backend/src/main/java/org/guvnor/common/services/project/backend/server/POMServiceImpl.java]: The path is the same, but the content has been updated. So the two modules are different. Old is what it was before update and the one loaded on line 241 has the latest content.

**Comment by Rikkola** [comment]
Did the fixes. Currently investigating if `RenameModuleEvent` should be updated to `UpdateModuleEvent`. Might be better to follow more than than just the name update. For example GAV changes should out date the `WorkbenchProjectContext` if the module with changes it the active module.

**Comment by Rikkola** [comment]
I'll still add the sub module update tests that @jhrcek suggested.

**Comment by Rikkola** [comment]
Yep. Moved to `UpdateModuleEvent`. If any location only cares about the name, it can compare the new and old module name.

File: uberfire-project/uberfire-project-backend/src/main/java/org/guvnor/common/services/project/backend/server/POMServiceImpl.java [hunk 4]
```diff
@@ -71,11 +79,17 @@ public POMServiceImpl(final @Named("ioStrategy") IOService ioService,
                           final POMContentHandler pomContentHandler,
                           final M2RepoService m2RepoService,
                           final MetadataService metadataService,
+                          final Event<ModuleUpdatedEvent> renameModuleEvent,
```
Comment (jhrcek) [inline uberfire-project/uberfire-project-backend/src/main/java/org/guvnor/common/services/project/backend/server/POMServiceImpl.java]: This param name is now out of sync (the instance field too).

File: uberfire-project/uberfire-project-backend/src/test/java/org/guvnor/common/services/project/backend/server/POMServiceImplSaveTest.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    private IOService ioService;
+
+    @Mock
+    private POMContentHandler pomContentHandler;
+
+    @Mock
+    private ModuleService moduleService;
+
+    @Mock
+    private EventSourceMock<ModuleUpdatedEvent> renameModuleEvent;
```
Comment (jhrcek) [inline uberfire-project/uberfire-project-backend/src/test/java/org/guvnor/common/services/project/backend/server/POMServiceImplSaveTest.java]: this should be renamed too..

File: uberfire-project/uberfire-project-backend/src/test/java/org/guvnor/common/services/project/backend/server/POMServiceImplSaveTest.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+                                     mock(PomEnhancer.class));
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        testFileSystem.tearDown();
+    }
+
+    @Test
+    public void whenProjectSavedWithoutRename_renameModuleEventNotFired() throws Exception {
```
Comment (jhrcek) [inline uberfire-project/uberfire-project-backend/src/test/java/org/guvnor/common/services/project/backend/server/POMServiceImplSaveTest.java]: And here and the test below have it in the name too :)

File: uberfire-project/uberfire-project-client/src/test/java/org/guvnor/common/services/project/client/context/WorkspaceProjectContextTest.java [hunk 7]
```diff
@@ -200,4 +202,50 @@ public void testContextChangeHandlerGetsRemoved() throws Exception {
                never()).onChange(any(),
                                  any());
     }
+
+    @Test
+    public void testNoRenameWhenNoActiveModule() throws Exception {
```
Comment (jhrcek) [inline uberfire-project/uberfire-project-client/src/test/java/org/guvnor/common/services/project/client/context/WorkspaceProjectContextTest.java]: and in this class too..

**Comment by jhrcek** [comment]
Also Jenkins execute full downstream build

File: uberfire-project/uberfire-project-backend/src/test/java/org/guvnor/common/services/project/backend/server/POMServiceImplSaveTest.java [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+                             "desctiption",
+                             "url",
+                             new GAV()),
+                     mock(Metadata.class),
+                     "");
+
+        verify(renameModuleEvent).fire(any());
+    }
+
+//    @Test
```
Comment (ederign) [inline uberfire-project/uberfire-project-backend/src/test/java/org/guvnor/common/services/project/backend/server/POMServiceImplSaveTest.java]: Please delete the commented code.

**Comment by Rikkola** [comment]
Ok the PR updated. Added tests for the submodule updates.

**Comment by jhrcek** [comment]
@Rikkola FYI I managed to build this locally, will provide feedback in the afternoon (after all the meetings :P )

**Comment by Rikkola** [comment]
@jhrcek No hurry from my side.

**Comment by jhrcek** [comment]
Tested this locally and after changing name in General settings and clicking Save, the name is correctly changed in breadcrumbs, but the title still remains the same as shown below:

titlenotupdated

**Comment by Rikkola** [comment]
I'll take a look at that.

**Comment by Rikkola** [comment]
ProjectScreen title fixed.

**Pull request merged**